pytorch_binding/test/new_benchmark.py

```diff
@@ -0,0 +1,158 @@
+    g = torch.randn(N, U+1, H, dtype=torch.float, device=0)
+    ys = torch.randint(1, V, (N, U), dtype=torch.int, device=0)
+
+    if rand_length:
+        lf = torch.randint(T // 2, T+1, (N,), dtype=torch.int, device=0)
+        lg = torch.randint(U // 2, U+1, (N,), dtype=torch.int, device=0)
+        lf = lf + T - lf.max()
+        lg = lg + U - lg.max()
+    else:
+        lf = torch.full((N, ), fill_value=T, dtype=torch.int, device=0)
+        lg = torch.full((N, ), fill_value=U, dtype=torch.int, device=0)
+
+    return f, g, ys, lf, lg
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Benchmark RNN-T loss implementation")
+    parser.add_argument("--loss", type=str, required=True,
+                        help="The target implementation")
+    parser.add_argument("--random-length", action="store_true", default=False,
+                        help="The random length")
+    parser.add_argument("--fwd-only", action="store_true", default=False,
+                        help="If True, benchmarking forward pass only, "
+                        "otherwise both forward & backward are benchmarked.")
+
+    args = parser.parse_args()
+    torch.manual_seed(0)
+
+    if args.loss == "warp-rnnt":
+        from warp_rnnt import rnnt_loss
+
+        def run_loss(xs, ys, xn, yn):
+            return rnnt_loss(xs, ys, xn, yn, gather=False)
+
+    elif args.loss == "warp-rnnt-gather":
+        from warp_rnnt import rnnt_loss
+
+        def run_loss(xs, ys, xn, yn):
+            return rnnt_loss(xs, ys, xn, yn, gather=True)
+
+    elif args.loss == "warp-rnnt-compact":
+        from warp_rnnt import rnnt_loss
+
+        def run_loss(xs, ys, xn, yn):
+            ys = torch.cat([ys[i, :yn[i]] for i in range(ys.size(0))])
+            return rnnt_loss(xs, ys, xn, yn, compact=True)
+    else:
+        raise ValueError(f"Unrecognized type of loss:{args.loss}")
+
+    H = 512
+    for E, T, U, V in [(10, 150, 40, 28), (10, 150, 20, 5000), (10, 1500, 300, 50)]:
+        for N in [1, 16, 32, 64, 128]:
+            joiner = Joint(H, V).cuda()
+            if args.loss.split('-')[-1] == 'compact':
+                joiner.requires_packing_()
+            f, g, ys, xn, yn = gen_data(
+                N, T, U, V, H,
+                rand_length=args.random_length
+            )
+            if args.fwd_only:
+                joiner.requires_grad_(False)
+            else:
+                f.requires_grad_(True)
+                g.requires_grad_(True)
+
+            print(
+                f"H={H:<5} T={T:<5} U={U:<5} V={V:<5} N={N:<5} shuffle-length={args.random_length}  fwd-only={args.fwd_only}")
+            with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA]) as prof:
+                with record_function(args.loss):
+                    for i in range(E):
+                        xs = joiner(f, g, xn, yn)
+                        loss = run_loss(xs, ys, xn, yn)
```

Comment:
I wonder at what stage cuda is synced. I suppose it's happened implicitly at the packing stage, when `xs` created.

Reply:
I'm not familiar with torch profiler, but I guess it sync at the enter and exit of the `with` context manager. I also tried
```
                with record_function(args.loss):
                    # torch.cuda.synchronize()
                    for i in range(E):
                        xs = joiner(f, g, xn, yn)
                        loss = run_loss(xs, ys, xn, yn)
                        if not args.fwd_only:
                            (loss.mean(dim=0)).backward()
                            joiner.zero_grad()
                            f.grad = None
                            g.grad = None
                    # torch.cuda.synchronize()
```
It seems these two lines of manually synchronizing make no difference, so the profiler probably has done the sync.